tests/unit/src/Internal/Service/DuplicatePaymentPreventionServiceTest.php

```diff
@@ -0,0 +1,480 @@
+		$attached_intent_id = 'pi_attached_intent_id';
+
+		// Arrange intent ID.
+		$this->mock_order_service
+			->expects( $this->once() )
+			->method( 'get_intent_id' )
+			->with( $this->processing_order_id )
+			->willReturn( $attached_intent_id );
+
+		// Arrange mock get_intention.
+		$attached_intent = WC_Helper_Intention::create_intention(
+			[
+				'id'       => $attached_intent_id,
+				'status'   => $intent_authorized_status,
+				'metadata' => [ 'order_id' => $this->processing_order_id ],
+			]
+		);
+
+		$this->mock_wcpay_request( Get_Intention::class, 1, $attached_intent_id )
+			->expects( $this->once() )
+			->method( 'format_response' )
+			->willReturn( $attached_intent );
+
+		// Act then assert.
+		$result = $this->sut->get_authorized_payment_intent_attached_to_order( $this->processing_order_id );
+		$this->assertSame( $attached_intent, $result );
+	}
+
+	public function provider_get_previous_paid_duplicate_order_id() {
+		// phpcs:disable
+		return [
+			'Expected null due to session ID not existed'   => [
+				null, 'Any session cart Hash', 888, true, // Session Order Data.
+				111, 'Any current cart Hash', 888, true,  // Current Order Data.
+				null,                                     // Expected value.
+			],
+			'Expected null due to same order_id'   => [
+				111, 'Any session cart Hash', 888, true,
+				111, 'Any current cart Hash', 888, true,
+				null,
+			],
+			'Expected null due to different cart hash'   => [
+				111, 'Any session cart Hash', 888, true,
+				222, 'Any current cart Hash', 888, true,
+				null,
+			],
+			'Expected null due to different customer ID'   => [
+				111, 'SAME_CART_HASH',  888, true,
+				222, 'SAME_CART_HASH',  999, true,
+				null,
+			],
+			'Expected null due to current order is NOT pending'   => [
+				111, 'SAME_CART_HASH', 999, true,
+				222, 'SAME_CART_HASH', 999, false,
+				null,
+			],
+			'Expected null due to session order is NOT paid'   => [
+				111, 'SAME_CART_HASH', 999, false,
+				222, 'SAME_CART_HASH', 999, true,
+				null,
+			],
+			'Expected session_order_id due to all conditions are matched'   => [
+				111, 'SAME_CART_HASH', 999, true,
+				222, 'SAME_CART_HASH', 999, true,
+				111
+			],
+		];
+		// phpcs:enable WordPress.Arrays.ArrayDeclarationSpacing.AssociativeArrayFound
+	}
+
+	/**
+	 * Note: Separate variable groups in the function signature to match with the provider data, for ease of understanding how they are grouped.
+	 *
+	 * @dataProvider provider_get_previous_paid_duplicate_order_id
+	 */
+	public function test_get_previous_paid_duplicate_order_id(
+		?int $session_order_id,
+		string $session_order_cart_hash,
+		int $session_order_customer_id,
+		bool $is_session_order_paid,
+
+		int $current_order_id,
+		string $current_order_cart_hash,
+		int $current_order_customer_id,
+		bool $is_current_order_pending,
+
+		?int $expected
+	) {
+		$this->sut = $this->getMockBuilder( DuplicatePaymentPreventionService::class )
```

Comment:
Let's add an inline comment here, defining the type correctly. Right now I'm getting errors in VSCode that DPPS methods are not defined.

Same applies to every instance of `$this->sut = $this->getMockBuilder` in the file.

```suggestion
		/**
		 * @var DuplicatePaymentPreventionService|MockObject
		 */
		$this->sut = $this->getMockBuilder( DuplicatePaymentPreventionService::class )
```

Reply:
Good call. I've created a separated $mocked_sut for tests using SUT with mocked methods.
621dd122fe1bb87df8a7c5677fe237e915896c08

Reply:
Thank you, but we still need to add doc blocks. VSCode is too smart about it, picking up the concrete type from the method: 

<img width="1063" alt="Screenshot 2023-10-26 at 8 22 42" src="https://github.com/Automattic/woocommerce-payments/assets/5311119/80a86af5-fa63-4d05-8b07-563f08a84133">

With that in mind, and given the fact that `mock_sut` is not defined within `setUp()`, I'd actually prefer to keep it as a local variable, instead of a property.

Reply:
> but we still need to add doc blocks. 

That's because I missed changing `sut` to `mocked_sut` for this test `test_clean_up_when_detecting_duplicate_order`. I've updated it e688822fc

I think it should be OK now. 

> given the fact that mock_sut is not defined within setUp(), I'd actually prefer to keep it as a local variable, instead of a property.

Compared with the benefit that we do not need to declare the mixed (MockObject and concrete type of the mock) type for every `$this->getMockBuilder()` call, IMO, this is acceptable.

Reply:
Also, `$this->mocked_sut` can help distinguish it from `$this->sut`. 

That said, I don't mind much if we change all current `$this->mocked_sut` back to `$this->sut` and declare the type of it. Let me know WDYT :D

Reply:
> Compared with the benefit that we do not need to declare the mixed (MockObject and concrete type of the mock) type for every $this->getMockBuilder() call, IMO, this is acceptable.

Take another look at the screenshot. I (also) didn't notice that `$this->sut` was used instead of `$this->mock_sut`, but the error is still applicable: VSCode does not care of the PHPDoc type of `$this->var`, and infers the type from the assignment earlier in the method. So we do need to define the type regardless, and with that in mind, the private property becomes useles..

Reply:
I think which IDE being used does matter here :D 

With the current code (declare the mix type for priate property `$this->mocked_sut`), PhpStorm can infer the correct type. 
![Markup on 2023-10-26 at 14:33:19](https://github.com/Automattic/woocommerce-payments/assets/10045087/f403ef8f-94d8-4794-94f7-84eb9ab8d645)

Without any declaration, just calling `$this->getMockBuilder`, PhpStorm can also infer the correct type: 

![Markup on 2023-10-26 at 14:37:27](https://github.com/Automattic/woocommerce-payments/assets/10045087/3f685af4-141a-4440-b217-dd02741f6b35)

--- 

Anyway, I am going to give it another try with your suggestion ff25309b639e7d465e44bc86e458861020def1af

If it's still not working for VSCode properly, might you submit a commit for it? :D

Reply:
> I think which IDE being used does matter here :D

It does. Seems like PhpStorm uses a combination of the property DocBlock and the assignment, joining them. In VSCode, the assignment takes priority, and the property is ignored.

Anyway, thank you for this! It works for me now.